packages/flutter/lib/src/widgets/routes.dart

```diff
@@ -1933,12 +2011,34 @@ abstract class ModalRoute<T> extends TransitionRoute<T> with LocalHistoryRoute<T
 
   @override
   void didChangeNext(Route<dynamic>? nextRoute) {
+    // In some cases, we need to get the delegated transition from a route before
+    // it is built, but some routes need to derive a delegated transition from
+    // the context. So if the next route does not have a `BuildContext` or a
+    // delegatedTransition, call nextRoute.setDelegatedTransition using the
+    // current `BuildContext`.
+    if (nextRoute is ModalRoute<T> && nextRoute.delegatedTransition == null && nextRoute.subtreeContext == null && this.subtreeContext != null) {
+      nextRoute.setDelegatedTransition(this.subtreeContext!);
+    }
+    if (nextRoute is ModalRoute<T> && canTransitionTo(nextRoute) && nextRoute.delegatedTransition.runtimeType != this.delegatedTransition.runtimeType) {
```

Comment:
The runtimeType checking feels iffy. In theory, this method should only be called when the next route changes, which seems to be rather rare. Why can we not accept whatever transition is given to us there? Also, builders may have changed even when the runtime type is the same. As shown in the example, it seems totally plausible that one would new up `VerticalDelegatedTransition`s with different builders.[^1]

Ideas from talking to Mitchell: 
- Maybe we need to fix something in the navigator to call this less frequently (only when the next route changes)
- For the route snapshotting: Maybe we need to move to a model where the Transition can look up whether to snapshot from the surrounding context instead of hard-coding that into the transition. Then, a snapshotting route could still donate a transition to another route without forcing it to snapshot (since from the context the receiving route would know not to snapshot).
[^1]: If we really need to do the runtime checking, `DelegatedTransition` should probably be designed to require you to override a method that implements the builder (so every instance of the subclass builds the same thing) instead of making it look as if you can pass in different builders.